### PR TITLE
Support ibc-go@v7.0.0

### LINKED ISSUE #7: Generic verifcation function support
ref. https://github.com/hyperledger-labs/yui-ibc-solidity/issues/165
https://github.com/datachainlab/ibc-mock-client/blob/main/modules/light-clients/xx-mock/types/client_state.go#L225
https://github.com/datachainlab/ibc-mock-client/blob/main/modules/light-clients/xx-mock/types/client_state.go#L251

The packet and ack commitment proof must be `sha256(value)` as with other state proofs.

### PR DESCRIPTION
**Pull request by siburu**
Also close #7

### CODE REVIEW

**Comment by siburu** [comment]
@bluele
Please review this PR.
The first commit mainly updates dependency management.
The second commit updates implementation details of mock client.

File: modules/light-clients/xx-mock/types/client_state.go [hunk 1]
```diff
@@ -47,299 +58,155 @@ func (cs ClientState) Status(_ sdk.Context, _ sdk.KVStore, _ codec.BinaryCodec)
...
-func (cs ClientState) GetProofSpecs() []*ics23.ProofSpec {
+// Validate performs a basic validation of the client state fields.
+func (cs ClientState) Validate() error {
 	return nil
 }
 
-// Validate performs basic validation of the client state fields.
-func (cs ClientState) Validate() error {
+// GetProofSpecs returns nil proof specs since Mock client doesn't use merkle proof.
+func (cs ClientState) GetProofSpecs() []*ics23.ProofSpec {
```
Comment (bluele) [inline modules/light-clients/xx-mock/types/client_state.go]: looks like GetProofSpecs was removed from ClientState interface

File: modules/light-clients/xx-mock/types/update.go [hunk 2]
```diff
@@ -1,38 +1,85 @@
...
+func (cs ClientState) UpdateState(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, clientMsg exported.ClientMessage) []exported.Height {
+	header, ok := clientMsg.(*Header)
+	if !ok {
+		panic(fmt.Errorf("expected type %T, got %T", &Header{}, clientMsg))
+	}
+
+	// check for duplicate update
+	if _, found := getConsensusState(clientStore, cdc, header.GetHeight()); found {
+		// perform no-op
+		return []exported.Height{header.GetHeight()}
```
Comment (bluele) [inline modules/light-clients/xx-mock/types/update.go:61]: should we return an empty slice instead here?

File: modules/light-clients/xx-mock/types/client_state.go [hunk 3]
```diff
@@ -47,299 +58,155 @@ func (cs ClientState) Status(_ sdk.Context, _ sdk.KVStore, _ codec.BinaryCodec)
...
-			channeltypes.ErrInvalidChannel,
-			"expected channel type %T, got %T", channeltypes.Channel{}, channelEnd)
+	if _, found := getConsensusState(clientStore, cdc, height); !found {
+		return sdkerrors.Wrap(clienttypes.ErrConsensusStateNotFound, "please ensure the proof was constructed against a height that exists on the client")
 	}
 
-	bz, err := cdc.Marshal(&channel)
-	if err != nil {
-		return err
+	if len(proof) != 0 {
```
Comment (bluele) [inline modules/light-clients/xx-mock/types/client_state.go:151]: NOTE: I understood that you followed solidity mock-client impl in https://github.com/hyperledger-labs/yui-ibc-solidity/blob/main/contracts/clients/MockClient.sol#L166

**Review by bluele — COMMENTED** [review]
Great! I left some comments. @siburu

File: modules/light-clients/xx-mock/types/client_state.go (hunk 1 above)
Comment (siburu) [reply modules/light-clients/xx-mock/types/client_state.go]: Oh, you are right. I will delete this. Thank you.

Funny thing is, ics-06 and ics-07 implementations still have `GetProofSpecs`, but I can't find where they are called.
Maybe it's a bug. I will submit a PR to remove them to ibc-go.

File: modules/light-clients/xx-mock/types/client_state.go (hunk 3 above)
Comment (siburu) [reply modules/light-clients/xx-mock/types/client_state.go:151]: Yes. Was it the right choice?

File: modules/light-clients/xx-mock/types/update.go (hunk 2 above)
Comment (siburu) [reply modules/light-clients/xx-mock/types/update.go:61]: I haven't been able to find the exact spec on this, but Tendermint client returns one height in the same case.
https://github.com/cosmos/ibc-go/blob/v7.0.0/modules/light-clients/07-tendermint/update.go#L141-L142

File: modules/light-clients/xx-mock/types/client_state.go (hunk 1 above)
Comment (bluele) [reply modules/light-clients/xx-mock/types/client_state.go]: In my understanding, ics-07 should refer to the proof specs to support store types other than IAVL. (e.g. SMT)

File: modules/light-clients/xx-mock/types/client_state.go (hunk 3 above)
Comment (bluele) [reply modules/light-clients/xx-mock/types/client_state.go:151]: Yes, you are right :)

File: modules/light-clients/xx-mock/types/client_state.go (hunk 1 above)
Comment (siburu) [reply modules/light-clients/xx-mock/types/client_state.go]: Yes, the `ProofSpecs` field of Tendermint client is still referred to in many places of the source code.
However, `GetProofSpecs` is not called from anywhere.

Comment (siburu) [reply modules/light-clients/xx-mock/types/client_state.go]: I deleted `GetProofSpecs` from Mock client.

File: modules/light-clients/xx-mock/types/update.go (hunk 2 above)
Comment (bluele) [reply modules/light-clients/xx-mock/types/update.go:61]: Yes, it is true, but I think it does not satisfy the following comment at least:
https://github.com/cosmos/ibc-go/blob/v7.0.0/modules/light-clients/07-tendermint/update.go#LL127C4-L127C63
> A list containing the updated consensus height is returned.

Comment (siburu) [reply modules/light-clients/xx-mock/types/update.go:61]: I see. ~I will ask about this in the ibc-go repo.~

**Review by bluele — APPROVED** [review]
LGTM👍
One comment about the returned heights from `UpdateState` remains, but the spec is unclear, so I approve this PR.

File: modules/light-clients/xx-mock/types/update.go (hunk 2 above)
Comment (siburu) [reply modules/light-clients/xx-mock/types/update.go:61]: I found [this conversation](https://github.com/cosmos/ibc-go/pull/1285/commits/9116fd127152e6fee2cc3a19c9807a6788e6db0e#r857602843) by going back through the conversation history on GitHub.

Comment (siburu) [reply modules/light-clients/xx-mock/types/update.go:61]: To sum up, heights returned by `UpdateState` are used for logging and events.
If the slice is empty at the time, there may be a problem.

Comment (siburu) [reply modules/light-clients/xx-mock/types/update.go:61]: @bluele As a conclusion, I'm thinking of keeping it as is.

**Comment by siburu** [comment]
@bluele Thank you! By the way, it seems that I can't merge PRs to this repo by myself.
Please merge this PR unless there remains a problem.

**Comment by bluele** [comment]
@siburu Merged and updated your permissions about the repo.

**Pull request merged**